### PR TITLE
Prevent scaling machine pools to 0 if it also scales down the only etcd or controlplane machine

### LINKED ISSUE #5454: [inconsistent options] `scale down` available in node `...` options, but not on `cluster` -> `machinePools` page
**Setup**
- Rancher version: v2.6.4-rc10
- Browser type & version: chrome

**Describe the bug**
options for a given node for scaling down are disabled when pool count = 1 and pool has worker label, but are enabled in the `...` options for that node. 

**To Reproduce**
* provision a cluster with 1 node per role
  * note that the `scale down` option is disabled for all nodes
* click the `...` option 
  * note that you can scale down the cluster's worker node. 
**Result**
* pool is removed/scaled down

**Expected Result**
Options for users should be consistent in the UI

**Screenshots**
Screen Shot 2022-03-21 at 3 39 45 PM
Screen Shot 2022-03-21 at 3 39 41 PM

### PR DESCRIPTION
**Pull request by mantis-toboggan-md**
### Summary
Fixes #5454 
The machine pool scale down button is disabled when the pool spec has only one replica defined; the individual pool scale down action menu option was disabled when the machine was the only instance of a controlplane or etcd role. This meant that if a cluster had one pool of workers with one worker machine, the pool couldn't be scaled down but the machine could. 

### Occurred changes and/or fixed issues
This PR updates the machine's `canScaleDown` function to prevent scaling down when it's the only machine in a pool. The role-checking logic is removed, since a machine can't be the only instance of a role without also being the only machine in its pool.

### CODE REVIEW

**Review by richard-cox — COMMENTED** [review]
I've requested some design clarifications from Gary over at https://github.com/rancher/dashboard/issues/5454#issuecomment-1104464507. If we go down this path some of the other locations will need to change to avoid other discrepancies between the three ways to scale

**Comment by richard-cox** [comment]
Additional comment, https://github.com/rancher/dashboard/issues/5454#issuecomment-1105025068. Waiting on clarification on slack regarding how scaling to zero should (or shouldn't) work

**Review by richard-cox — CHANGES_REQUESTED** [review]
Ok, think it's sorted now, see https://github.com/rancher/dashboard/issues/5454#issuecomment-1106688959

**Comment by mantis-toboggan-md** [comment]
Updated the +/- buttons to use (essentially) the same logic as the individual machine action menu item: the - button should now be disabled when scaling down the pool would result in the loss of the only etcd or controlplane machine in the cluster.

File: models/cluster.x-k8s.io.machine.js [hunk 1]
```diff
@@ -7,7 +7,44 @@ import { insertAt, findBy } from '@/utils/array';
 import { get } from '@/utils/object';
 import { downloadUrl } from '@/utils/download';
 import SteveModel from '@/plugins/steve/steve-class';
+// Prevent scaling down control plane or etcd nodes to zero
+// This is a little overly optimised but avoids iterating over the whole machine set every time
```
Comment (rak-phillip) [inline models/cluster.x-k8s.io.machine.js:11]: Personally, I would prefer JSDoc style comments to provide a little more context about our parameters and what we're returning. I find the function name of `noOnlyOfRole` to be a bit unclear, especially when `type` is used so much throughout the function body.

**Review by rak-phillip — COMMENTED** [review]
I might have some reservations in that I can scale down `etcd` nodes rapidly enough that I can put my cluster in a failure state.

image

image

I'm able to get into this state by selecting the overflow menu => scale down all of the machines in a pool, one after the other. Perhaps, another set of eyes might not be a bad idea.

**Review by richard-cox — APPROVED** [review]
@mantis-toboggan-md I've made some small tweaks and tested (see commit message). Unless there's any objections I'll merge tomorrow.

I'd agree with Phil's comment regarding scaling down to bad states, I think you/he encountered issues as per Brandon's comment. Under the hood we set an annotation on each effected machine and then scale down the pool the machines are in by that amount. Smarts for doing this correctly might need to come from the backend. In terms of scaling by the pool buttons, it's possible to hit the same bad state, but requires more effort.

Given that this PR addresses bottom line machine vs type availability I'd say we should still get this in. Further improvements (both frontend and backend) came come in a later one.

**Pull request merged**